lib/nerves/artifact/resolvers/github_api.ex

```diff
@@ -2,143 +2,132 @@ defmodule Nerves.Artifact.Resolvers.GithubAPI do
+    token = System.get_env("GITHUB_TOKEN") || System.get_env("GH_TOKEN")
 
-    case Enum.split_with(@required_opts, &(&1 in keys)) do
-      {_, [_ | _] = missing} -> {:error, missing}
-      _ -> {:ok, opts}
+    if token do
+      # Let the env var take precedence
+      %{opts | token: token}
+    else
+      opts
     end
   end
 
-  defp get_asset_url(assets, artifact_name) do
-    ret =
-      Enum.find(assets, fn %{"name" => name} ->
-        String.equivalent?(artifact_name, name)
-      end)
+  defp maybe_start_http(%{http_pid: pid} = opts) when is_pid(pid), do: opts
 
-    case ret do
-      nil -> {:error, "No artifact found"}
-      %{"url" => url} -> {:ok, url}
-    end
+  defp maybe_start_http(opts) do
+    {:ok, http_pid} = opts.http_client.start_link()
+    %{opts | http_pid: http_pid}
   end
 
-  defp get_username(opts, artifact_name) do
-    case Keyword.get(opts, :username) do
-      nil ->
-        Mix.raise("""
-        GitHub username not set for artifact: #{artifact_name}.
-
-        Ensure that you have your GitHub username set correctly
-        in your environment. You might need to export an environmental
-        variable.
-
-        For example:
-
-        export GITHUB_USER=<your_username>
+  defp fetch_artifact(opts) do
+    info = if System.get_env("NERVES_DEBUG") == "1", do: opts.url, else: opts.artifact_name
 
-        For correctly setting up your environment please see the documentation for the artifact you are
-        trying to download.
-        """)
+    Utils.Shell.info(["  [GitHub] ", info])
 
-      username ->
-        username
+    with {:ok, data} <- release_details(opts),
+         %{"assets" => assets} <- Utils.json_decode(data),
+         {:ok, asset_url} <- get_asset_url(assets, opts) do
+      opts.http_client.get(opts.http_pid, asset_url,
+        headers: [{"Accept", "application/octet-stream"} | opts.headers]
+      )
     end
   end
 
-  defp get_token(opts, artifact_name) do
-    case Keyword.get(opts, :token) do
-      nil ->
-        Mix.raise("""
-        GitHub token not set for artifact: #{artifact_name}
+  defp release_details(opts) do
+    case opts.http_client.get(opts.http_pid, opts.url, headers: opts.headers, progress?: false) do
+      {:error, "Status 404 Not Found"} ->
+        invalid_token? = is_nil(opts.token) or opts.token == ""
 
-        Ensure that you have your GitHub token set correctly
-        in your environment. You might need to export an environmental
-        variable.
+        msg =
+          if not opts.public? and invalid_token? do
+            """
+            Missing token
 
-        For example:
+                 For private releases, you must authenticate the request to fetch release assets.
+                 You can do this in a few ways:
 
-        export GITHUB_TOKEN=<your_token>
+                   * export or set GITHUB_TOKEN=<your-token>
+                   * set `token: <your-token>` for this artifact site in your Nerves system mix.exs
```

Comment:
```suggestion
                   * set `token: <get-token-function>` for this GitHub repository in your Nerves system mix.exs
```

Since we don't want people storing tokens in their `mix.exs`, I changed the line to give a hint to use a function. 

I also changed artifact site to GitHub repository, since I think that more directly corresponds to what they need from GitHub. There might be a better word?

Reply:
Ya, I was struggling on how best to word this to direct users to their `nerves_system -> mix.exs -> nerves_package/0 -> :artifact_sites` setting. I think this change is better